### PR TITLE
[SYCL-MLIR]: Memory access analysis (part5)

### PR DESCRIPTION
**Pull request by etiotto**
Extend the memory access analysis to handle memory accesses indexed by SYCL global thread ids (in addition to loop IVs).

### CODE REVIEW

File: polygeist/lib/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.cpp [hunk 1]
```diff
@@ -100,12 +98,26 @@ template <typename T> static SetVector<T> getParentsOfType(Block *block) {
   return res;
 }
 
+/// Retrieve the operations with the specified type in the given function.
+template <typename T>
+static SetVector<T> getOperationsOfType(FunctionOpInterface funcOp) {
+  SetVector<T> res;
+  for (Block &block : funcOp.getFunctionBody()) {
```
Comment (whitneywhtsang) [inline polygeist/lib/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.cpp:105]: use `walk` to make it recursive, as operations can be nested.

File: polygeist/lib/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.cpp [hunk 2]
```diff
@@ -325,9 +336,14 @@ static ValueOrMultiplier getMultiplier(const Value expr, const Value factor,
...
       })
-      .Case<arith::ExtUIOp, arith::ExtSIOp, arith::TruncIOp, arith::IndexCastOp,
-            arith::IndexCastUIOp>([&](auto castOp) {
-        return getMultiplier(castOp.getOperand(), factor, solver);
+      .Case<sycl::SYCLNDItemGetGlobalIDOp>(
+          [&](auto getGlobalIdOp) -> ValueOrMultiplier {
+            if (getGlobalIdOp.getResult() == factor)
+              return Multiplier::one(expr.getContext());
+            return Value();
+          })
```
Comment (whitneywhtsang) [inline polygeist/lib/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.cpp:344]: if we remove `if (!op) {` from line 250, do we still need this new case?

**Review by whitneywhtsang — APPROVED** [review]

File: polygeist/lib/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.cpp [hunk 3]
```diff
@@ -1007,44 +1023,82 @@ void MemoryAccessAnalysis::build(T memoryOp, DataFlowSolver &solver) {
...
+    llvm::dbgs() << "\n";
   });
 
-  // Construct the memory access matrix. The number of rows is equal to the
-  // dimensionality of the sycl.id used by the accessor subscript operation. The
-  // number of columns is equal to the number of loops surrounding the memory
-  // access plus (TODO) the dimensionality of the sycl item received by the
-  // kernel.
+  // Collect the "get_global_ids" operations (yielding the global thread ids).
+  std::vector<sycl::SYCLNDItemGetGlobalIDOp> getGlobalIdOps =
```
[the remark this answers is not in the document]
Comment (sommerlukas) [reply polygeist/lib/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.cpp:1031]: Ok, makes sense to do this in a separate PR then 👍

File: polygeist/lib/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.cpp [hunk 4]
```diff
@@ -1007,44 +1023,82 @@ void MemoryAccessAnalysis::build(T memoryOp, DataFlowSolver &solver) {
...
+  for (sycl::SYCLNDItemGetGlobalIDOp getGlobalIdOp : getGlobalIdOps)
+    loopAndThreadVars.emplace_back(getGlobalIdOp.getResult());
   for (AffineForOp loop : llvm::reverse(enclosingLoops))
-    IVs.push_back(loop.getInductionVar());
+    loopAndThreadVars.emplace_back(loop.getInductionVar());
+
+  // Construct the memory access matrix. The number of rows is equal to the
+  // dimensionality of the sycl.id used by the accessor subscript operation.
+  // The number of columns is equal to the number of loops surrounding the
+  // memory access plus the number of threads used in the kernel.
```
[the remark this answers is not in the document]
Comment (sommerlukas) [reply polygeist/lib/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.cpp]: Ok, thanks. 

"Number of dimensions of the thread ID" would be even more specific, but both are fine with me.

**Review by sommerlukas — APPROVED** [review]

**Pull request merged**